secondarylib-feedback/src/main/scala/com/socrata/datacoordinator/secondary/feedback/FeedbackContext.scala

```diff
@@ -0,0 +1,285 @@
+
+  def mergeWith[A, B](xs: Map[A, B], ys: Map[A, B])(f: (B, B) => B): Map[A, B] = {
+    ys.foldLeft(xs) { (combined, yab) =>
+      val (a, yb) = yab
+      val newB = combined.get(a) match {
+        case None => yb
+        case Some(xb) => f(xb, yb)
+      }
+      combined.updated(a, newB)
+    }
+  }
+
+  // this may throw a ComputationFailure exception
+  private def computeUpdates(computationHandler: ComputationHandler[CT, CV], rows: IndexedSeq[Row[CV]]): Map[Int, Map[UserColumnId, CV]] = {
+    val perDatasetData = computationHandler.setupDataset(cookie)
+    val perColumnData = cookie.strategyMap.toSeq.collect {
+      case (targetCol, strat) if computationHandler.matchesStrategyType(strat.strategyType) =>
+        computationHandler.setupColumn(perDatasetData, strat, targetCol)
+    }
+    val toCompute = rows.iterator.zipWithIndex.flatMap { case (row, index) =>
+      val rcis =
+        perColumnData.flatMap { columnData =>
+          // don't compute if there has been to change to the source columns
+          if (noChange(row, columnData.strategy.sourceColumnIds))
+            None
+          else
+            Some(computationHandler.setupCell(columnData, row.data))
+        }
+      if (rcis.isEmpty) Iterator.empty
+      else Iterator.single(index -> rcis)
+    }.toMap
+
+    val results = computationHandler.compute(toCompute)
+
+    results.flatMap { case (rowIdx, updates) =>
+      val row = rows(rowIdx)
+      val filtered = updates.filter { case (colId: UserColumnId, value) =>
+        extractCV(row.data, colId) != Some(value) // don't update to current value
+      }
+      val rowId = extractCV(row.data, cookie.primaryKey) match {
+        case None => throw new Exception(s"Cannot find value for primary key ${cookie.primaryKey} in row: $row!")
+        case Some(other) => other
+      }
+      if (filtered.nonEmpty) {
+        val newRow = Map(
+          cookie.primaryKey -> rowId
+        ) ++ filtered.map { case (id, value) => (id, value)}.toMap
+        Some(rowIdx -> newRow)
+      } else {
+        None
+      }
+    }
+  }
+
+  private def noChange(row: Row[CV], columns: Seq[UserColumnId]): Boolean = row.oldData match {
+    case Some(old) =>
+      columns.forall { id =>
+        val internal = cookie.columnIdMap(id)
+        row.data(internal) == old(internal) // safe because updates contain _all_ row values (and this must be one)
+      }
+    case None => false
+  }
+
+  private def extractCV(row: secondary.Row[CV], userId: UserColumnId): Option[CV] = {
+    val internal = cookie.columnIdMap(userId)
+    row.get(internal)
+  }
+
+  // commands for mutation scripts
+  val commands = Seq(
+    JObject(Map(
+      "c" -> JString("normal"),
+      "user" -> JString(user)
+    )),
+    JObject(Map(
+      "c" -> JString("row data"),
+      "update_only" -> JBoolean.canonicalTrue,
+      "nonfatal_row_errors" -> JArray(Array(JString("insert_in_update_only")))
+    ))
+  )
```

Comment:
Generally, with modern rojoma-json, it's better to use the json interpolator than building objects by hand.  Much nicer to read, and just as safe (or unsafe..)